File: cellular_automata/cpp/cellular/cellular.tests.cpp
#include "cellular.hpp"

#include <catch2/catch_test_macros.hpp>
#include <catch2/generators/catch_generators.hpp>

TEST_CASE("next() executes rule") {
  SECTION("Rule 0 makes everything empty") {

    const auto a = GENERATE(ca::EMPTY, ca::FILLED);
    const auto b = GENERATE(ca::EMPTY, ca::FILLED);
    const auto c = GENERATE(ca::EMPTY, ca::FILLED);

    SECTION(std::string{ a, b, c }) {
      REQUIRE(ca::State{ ca::EMPTY, ca::EMPTY, ca::EMPTY } == ca::next({ a, b, c }, 0));
    }
  }
}

File: cellular_automata/python/cellular/cellular_tests.py
import unittest
from cellular import cellular


class CellularTests(unittest.TestCase):

    def test_rule_0_makes_everything_empty(self):
        for a in [cellular.EMPTY, cellular.FILLED]:
            for b in [cellular.EMPTY, cellular.FILLED]:
                for c in [cellular.EMPTY, cellular.FILLED]:
                    with self.subTest(msg=f"{a}{b}{c}"):
                        self.assertEqual("".join([cellular.EMPTY, cellular.EMPTY, cellular.EMPTY]),
                                         cellular.next_state("".join([a, b, c]), 0))


if __name__ == '__main__':
    unittest.main()
